resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddressStreamProviders.java

```diff
@@ -18,18 +18,43 @@
 import io.netty.util.internal.PlatformDependent;
 import io.netty.util.internal.UnstableApi;
 
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
 /**
  * Utility methods related to {@link DnsServerAddressStreamProvider}.
  */
 @UnstableApi
 public final class DnsServerAddressStreamProviders {
+    // We use 5 minutes which is the same as what OpenJDK is using in sun.net.dns.ResolverConfigurationImpl.
+    private static final long REFRESH_INTERVAL = TimeUnit.MINUTES.toNanos(5);
+
     // TODO(scott): how is this done on Windows? This may require a JNI call to GetNetworkParams
     // https://msdn.microsoft.com/en-us/library/aa365968(VS.85).aspx.
     private static final DnsServerAddressStreamProvider DEFAULT_DNS_SERVER_ADDRESS_STREAM_PROVIDER =
+            new DnsServerAddressStreamProvider() {
+        private volatile DnsServerAddressStreamProvider currentProvider;
+        private final AtomicLong lastRefresh = new AtomicLong(System.nanoTime());
+
+        @Override
+        public DnsServerAddressStream nameServerAddressStream(String hostname) {
+            long last = lastRefresh.get();
+            DnsServerAddressStreamProvider current = currentProvider;
+            if (System.nanoTime() - last > REFRESH_INTERVAL) {
+                if (lastRefresh.compareAndSet(last, System.nanoTime())) {
```

Comment:
This is racy. That's probably fine in the normal case, but it would at least deserve a comment that we may end up using the old configuration.

Note that the race means it can't be relied on for initializing `currentProvider` to be non-null.

Reply:
yeah I know its a racy but this is fine as it should only be for a small amount of time.